src/kv/map.h

```diff
@@ -2,723 +2,202 @@
+          {
+            // Deletions are indicated by nullopt. We cannot deserialise them,
+            // they are deletions here as well
+            typed_w[KSerialiser::from_serialised(uk)] = std::nullopt;
+          }
+          else
+          {
+            typed_w[KSerialiser::from_serialised(uk)] =
+              VSerialiser::from_serialised(opt_uv.value());
+          }
+        }
 
-    void clear() override
-    {
-      // This discards all entries in the roll and resets the compacted value
-      // and rollback counter. The Map expects to be locked before clearing it.
-      roll->clear();
-      roll->insert_back(create_new_local_commit(0, State(), Write()));
-      rollback_counter = 0;
+        hook(v, typed_w);
+      };
     }
 
-    void lock() override
+    void set_local_hook(const CommitHook& hook)
     {
-      sl.lock();
+      untyped_map.set_local_hook(wrap_commit_hook(hook));
     }
 
-    void unlock() override
+    void unset_local_hook()
     {
-      sl.unlock();
+      untyped_map.unset_local_hook();
     }
 
-    void swap(AbstractMap* map_) override
+    void set_global_hook(const CommitHook& hook)
     {
-      This* map = dynamic_cast<This*>(map_);
-      if (map == nullptr)
-        throw std::logic_error(
-          "Attempted to swap maps with incompatible types");
-
-      std::swap(rollback_counter, map->rollback_counter);
-      std::swap(roll, map->roll);
+      untyped_map.set_global_hook(wrap_commit_hook(hook));
     }
 
-    template <typename TView>
-    TView* create_view(Version version)
+    void unset_global_hook()
     {
-      lock();
-
-      // Find the last entry committed at or before this version.
-      TView* view = nullptr;
-
-      for (auto current = roll->get_tail(); current != nullptr;
-           current = current->prev)
-      {
-        if (current->version <= version)
-        {
-          view = new TView(
-            *this,
-            rollback_counter,
-            current->state,
-            roll->get_head()->state,
-            current->version);
-          break;
-        }
-      }
-
-      if (view == nullptr)
-      {
-        view = new TView(
-          *this,
-          rollback_counter,
-          roll->get_head()->state,
-          roll->get_head()->state,
-          roll->get_head()->version);
-      }
-
-      unlock();
-      return view;
+      untyped_map.unset_global_hook();
     }
   };
+
+  template <typename K, typename V>
+  using JsonSerialisedMap = TypedMap<
+    K,
+    V,
+    kv::serialisers::JsonSerialiser<K>,
+    kv::serialisers::JsonSerialiser<V>>;
+
+  template <typename K, typename V>
+  using MsgPackSerialisedMap = TypedMap<
+    K,
+    V,
+    kv::serialisers::MsgPackSerialiser<K>,
+    kv::serialisers::MsgPackSerialiser<V>>;
+
+  // The default kv::Map will use msgpack serialisers. Custom types are
+  // supported through the MSGPACK_DEFINE macro
+  template <typename K, typename V>
+  using Map = MsgPackSerialisedMap<K, V>;
```

Comment:
Specifying serialisers is slightly verbose, so we name some sensible defaults for you. And finally we provide `kv::Map<K, V>` again here. Note that this is a slight breaking change - anyone who had a custom hasher (ie `kv::Map<K, V, H>`) will get a compile error, and they'll have to remove it. We had no instances of this, and user custom types should no longer need it.